tests/densify/test_interface_patchmatchnet.py

```diff
@@ -0,0 +1,149 @@
+
+        self._num_tracks = DEFAULT_NUM_TRACKS
+        for j in range(self._num_tracks):
+            world_x = DEFAULT_TRACK_POINTS[j]
+            track_to_add = SfmTrack(world_x)
+
+            for i in range(self._num_valid_cameras):
+                measurement = self._sfm_result.get_camera(i).project(world_x)
+                track_to_add.add_measurement(idx=i, m=measurement)
+            self._sfm_result.add_track(track_to_add)
+
+    def test_dataset_length(self) -> None:
+        """Test whether the dataset length is equal to the number of valid cameras, because every valid camera will be
+        regarded as reference view once."""
+        self.assertEqual(len(self._dataset_patchmatchnet), self._num_valid_cameras)
+
+    def test_select_src_views(self) -> None:
+        """Test whether the (ref_view, src_view) pairs are selected correctly."""
+        self.assertTrue(
+            self._dataset_patchmatchnet.get_packed_pairs()
+            == [
+                {"ref_id": 0, "src_ids": [7, 1, 6, 2]},
+                {"ref_id": 1, "src_ids": [0, 2, 3, 7]},
+                {"ref_id": 2, "src_ids": [3, 1, 4, 0]},
+                {"ref_id": 3, "src_ids": [4, 2, 5, 1]},
+                {"ref_id": 4, "src_ids": [5, 3, 6, 2]},
+                {"ref_id": 5, "src_ids": [6, 4, 7, 3]},
+                {"ref_id": 6, "src_ids": [5, 7, 4, 0]},
+                {"ref_id": 7, "src_ids": [6, 0, 5, 1]},
+            ]
+        )
+
+    def test_depth_ranges(self) -> None:
+        """Test whether the depth ranges for each camera are calculated correctly and whether the depth outliers
+        (one too close and one too far) are filtered out in the depth range"""
+
+        # test the lower bound of the depth range
+        self.assertAlmostEqual(self._dataset_patchmatchnet._depth_ranges[EXAMPLE_CAMERA_ID][0], 10.60262, 2)
+        # test the upper bound of the depth range
+        self.assertAlmostEqual(self._dataset_patchmatchnet._depth_ranges[EXAMPLE_CAMERA_ID][1], 34.12857, 2)
+
+    def test_get_item(self) -> None:
+        """Test get item method when yielding test data from dataset."""
+        example = self._dataset_patchmatchnet[EXAMPLE_CAMERA_ID]
+
+        B, C, H, W = example["imgs"]["stage_0"].shape
+
+        # the batch size here means the number of views (1 ref view and (B-1) src views) used in inference.
+        #   the number of views must not be larger than the number of valid cameras, or the number of images
+        self.assertLessEqual(B, DEFAULT_NUM_IMAGES)
+
+        # test images size and channel number are correct
+        h, w, c = self._img_dict[0].value_array.shape
+        self.assertEqual((H, W, C), (h, w, c))
+
+        # test 3x4 projection matrices are correct
+        sample_proj_mat = example["proj_matrices"]["stage_0"][0][:3, :4]
+        sample_camera = self._sfm_result.get_camera(EXAMPLE_CAMERA_ID)
+        actual_proj_mat = sample_camera.calibration().K() @ sample_camera.pose().inverse().matrix()[:3, :4]
```

Comment:
can we just use the accurate number, and document which scene/image it corresponds to?

Reply:
marking as resolved, thanks